client/app/components/dashboards/AddWidgetDialog.jsx

```diff
@@ -0,0 +1,448 @@
+                        </a>
+                      ))}
+                    </div>
+                  }
+                </div>
+              }
+
+              {
+                (this.state.searchTerm !== '') &&
+                <div>
+                  {
+                    (this.state.searchedQueries.length === 0) &&
+                    <div className="text-muted">No results matching search term.</div>
+                  }
+                  {
+                    (this.state.searchedQueries.length > 0) &&
+                    <div className="list-group">
+                      {this.state.searchedQueries.map(query => (
+                        <a
+                          href="javascript:void(0)"
+                          className="list-group-item"
+                          key={query.id}
+                          onClick={() => this.selectQuery(query.id)}
+                          // eslint-disable-next-line react/no-danger
+                          dangerouslySetInnerHTML={{ __html: highlight(query.name, this.state.searchTerm) }}
+                        />
+                      ))}
+                    </div>
+                  }
+                </div>
+              }
+            </div>
+          }
+
+          {
+            this.state.selectedQuery &&
+            <div>
+              <div className="form-group">
+                <label>Choose Visualization</label>
+                <select
+                  className="form-control"
+                  onChange={event => this.selectVisualization(this.state.selectedQuery, event.target.value)}
+                >
+                  {visualizationGroups.map(visualizations => (
+                    <optgroup label={visualizations[0].type} key={visualizations[0].type}>
+                      {visualizations.map(visualization => (
+                        <option value={visualization.id} key={visualization.id}>{visualization.name}</option>
+                      ))}
+                    </optgroup>
+                  ))}
+                </select>
+              </div>
+            </div>
+          }
+
+          {
+            (this.state.parameterMappings.length > 0) &&
+            <div>
+              <label>Parameters</label>
+              {this.state.parameterMappings.map((mapping, index) =>
+                this.renderMapping(mapping, existingParamNames, index === 0))}
+            </div>
+          }
+        </div>
+
+        <div className="modal-footer">
+          <button
+            type="button"
+            className="btn btn-default"
+            disabled={this.state.saveInProgress}
+            onClick={this.props.dismiss}
+          >
+            Close
+          </button>
+          <button
+            type="button"
+            className="btn btn-primary"
+            disabled={this.state.saveInProgress}
+            onClick={() => this.saveWidget()}
+          >
+            Add to Dashboard
+          </button>
+        </div>
+      </div>
```

Comment:
Let's split this long render method into multiple components, one for each stage?

Reply:
Added benefit from splitting this to components: we should be able to reuse the component of the last phase (when you selected a query and visualization already) in the "Add to Dashboard" dialog on query page.